File: python-examples/17-order-processing/workflows.py
"""Order processing workflow with structured output and fraud review."""

from pydantic import BaseModel
from polos import workflow, WorkflowContext
from agents import order_agent
from schemas import OrderAgentOutput


class OrderPayload(BaseModel):
    """Input for order processing."""

    order_id: str
    customer_id: str
    customer_email: str
    amount: float


class OrderResult(BaseModel):
    """Result of order processing."""

    order_id: str
    status: str
    charge_id: str | None = None
    fraud_review_required: bool = False
    fraud_approved: bool | None = None
    email_sent: bool = False


@workflow(id="order_processing_workflow")
async def order_processing_workflow(
    ctx: WorkflowContext, payload: OrderPayload
) -> OrderResult:
    """Process an order with structured output and conditional fraud review.

    Flow:
    1. Agent charges customer via Stripe
    2. If amount > $1000: Agent requests fraud review, workflow suspends
    3. After fraud approval: Agent sends confirmation email
    4. If amount <= $1000: Agent sends confirmation immediately
    """
    # Step 1: Process the payment
    result = await ctx.step.agent_invoke_and_wait(
        "start_order",
        order_agent.with_input(
            f"Process payment for order {payload.order_id}. "
            f"Charge customer {payload.customer_id} for ${payload.amount:.2f} USD. "
            f"Customer email is {payload.customer_email}.",
        )
    )

    output: OrderAgentOutput = result.result
    charge_id = output.action_details.charge_id

    # Step 2: Check if fraud review is requested
    if output.action_requested == "fraud_review":
        execution_id = ctx.root_execution_id or ctx.execution_id
        await ctx.step.run("log_fraud_review", log_fraud_review, execution_id=execution_id)

        # Suspend for human fraud review
        resume_data = await ctx.step.suspend(
            "fraud_review",
            data={
                "order_id": payload.order_id,
                "customer_id": payload.customer_id,
                "amount": payload.amount,
                "charge_id": charge_id,
                "message": "Please review this order for fraud (amount > $1000)",
            },
            timeout=86400,
        )

        fraud_approved = resume_data.get("data", {}).get("approved", False)

        if not fraud_approved:
            print("\n--- Order rejected by fraud review ---")
            return OrderResult(
                order_id=payload.order_id,
                status="rejected",
                charge_id=charge_id,
                fraud_review_required=True,
                fraud_approved=False,
                email_sent=False,
            )

        # Step 3: Fraud approved - tell agent to send confirmation
        result = await ctx.step.agent_invoke_and_wait(
            "send_confirmation",
            order_agent.with_input(
                f"Fraud review APPROVED for order {payload.order_id}. "
                f"Charge ID: {charge_id}, Amount: ${payload.amount:.2f}. "
                f"Now send confirmation email to {payload.customer_email}.",
            )
        )

        output = result.result
        print(f"Agent action: {output.action}")
        print(f"Status: {output.status_message}")

        return OrderResult(
            order_id=payload.order_id,
            status="completed",
            charge_id=charge_id,
            fraud_review_required=True,
            fraud_approved=True,
            email_sent=True,
        )

    # No fraud review needed - agent already sent confirmation
    print("\n--- Order completed (no fraud review needed) ---")
    return OrderResult(
        order_id=payload.order_id,
        status="completed",
        charge_id=charge_id,
        fraud_review_required=False,
        fraud_approved=None,
        email_sent=True,
    )

async def log_fraud_review(execution_id: str):
    print("\n" + "*" * 60)
    print("*** FRAUD REVIEW REQUIRED ***")
    print("*** To approve: python approve_order.py " + execution_id)
    print("*** To reject:  python approve_order.py " + execution_id + " --reject")
    print("*" * 60 + "\n")

File: typescript-examples/17-order-processing/workflows.ts
/**
 * Order processing workflow with structured output and fraud review.
 *
 * Demonstrates:
 * - Agent invocation with structured output
 * - Conditional fraud review (suspend/resume) for high-value orders
 * - Human-in-the-loop pattern
 *
 * Flow:
 *   Amount <= $1000: Agent charges + sends confirmation immediately
 *   Amount > $1000:  Agent charges, requests fraud review, workflow suspends,
 *                    then sends confirmation on approval
 */

import { defineWorkflow } from '@polos/sdk';
import { orderAgent } from './agents.js';
import type { OrderAgentOutput } from './schemas.js';

// ============================================================================
// Payload / Result Types
// ============================================================================

export interface OrderPayload {
  orderId: string;
  customerId: string;
  customerEmail: string;
  amount: number;
}

export interface OrderResult {
  orderId: string;
  status: string;
  chargeId: string | null;
  fraudReviewRequired: boolean;
  fraudApproved: boolean | null;
  emailSent: boolean;
}

// ============================================================================
// Order Processing Workflow
// ============================================================================

export const orderProcessingWorkflow = defineWorkflow<OrderPayload, unknown, OrderResult>(
  { id: 'order_processing_workflow' },
  async (ctx, payload) => {
    // Step 1: Process the payment
    const result = (await ctx.step.agentInvokeAndWait(
      'start_order',
      orderAgent.withInput(
        `Process payment for order ${payload.orderId}. ` +
        `Charge customer ${payload.customerId} for $${payload.amount.toFixed(2)} USD. ` +
        `Customer email is ${payload.customerEmail}.`,
      ),
    )) as Record<string, unknown>;

    const output = result['result'] as OrderAgentOutput;
    const chargeId = output.action_details.charge_id;

    // Step 2: Check if fraud review is requested
    if (output.action_requested === 'fraud_review') {
      const executionId = ctx.rootExecutionId ?? ctx.executionId;

      await ctx.step.run(
        'log_fraud_review',
        () => {
          console.log('\n' + '*'.repeat(60));
          console.log('*** FRAUD REVIEW REQUIRED ***');
          console.log('*** To approve: npx tsx approve_order.ts ' + executionId);
          console.log('*** To reject:  npx tsx approve_order.ts ' + executionId + ' --reject');
          console.log('*'.repeat(60) + '\n');
        },
      );

      // Suspend for human fraud review
      const resumeData = await ctx.step.suspend<Record<string, unknown>, Record<string, unknown>>(
        'fraud_review',
        {
          data: {
            order_id: payload.orderId,
            customer_id: payload.customerId,
            amount: payload.amount,
            charge_id: chargeId,
            message: 'Please review this order for fraud (amount > $1000)',
          },
          timeout: 86400,
        },
      );

      const fraudApproved = Boolean(
        (resumeData?.['data'] as Record<string, unknown> | undefined)?.['approved'] ?? false,
      );

      if (!fraudApproved) {
        console.log('\n--- Order rejected by fraud review ---');
        return {
          orderId: payload.orderId,
          status: 'rejected',
          chargeId,
          fraudReviewRequired: true,
          fraudApproved: false,
          emailSent: false,
        };
      }

      // Step 3: Fraud approved - tell agent to send confirmation
      const confirmResult = (await ctx.step.agentInvokeAndWait(
        'send_confirmation',
        orderAgent.withInput(
          `Fraud review APPROVED for order ${payload.orderId}. ` +
          `Charge ID: ${chargeId ?? 'unknown'}, Amount: $${payload.amount.toFixed(2)}. ` +
          `Now send confirmation email to ${payload.customerEmail}.`,
        ),
      )) as Record<string, unknown>;

      const confirmOutput = confirmResult['result'] as OrderAgentOutput;
      console.log(`Agent action: ${confirmOutput.action}`);
      console.log(`Status: ${confirmOutput.status_message}`);

      return {
        orderId: payload.orderId,
        status: 'completed',
        chargeId,
        fraudReviewRequired: true,
        fraudApproved: true,
        emailSent: true,
      };
    }

    // No fraud review needed - agent already sent confirmation
    console.log('\n--- Order completed (no fraud review needed) ---');
    return {
      orderId: payload.orderId,
      status: 'completed',
      chargeId,
      fraudReviewRequired: false,
      fraudApproved: null,
      emailSent: true,
    };
  },
);
